crates/wasi-nn/src/tf.rs

```diff
@@ -0,0 +1,160 @@
+//! Implements the wasi-nn API.
+use crate::api::{Backend, BackendError, BackendExecutionContext, BackendGraph};
+use crate::witx::types::{ExecutionTarget, GraphBuilderArray, Tensor};
+use std::ops::Deref;
+use std::path::Path;
+use std::str;
+use std::sync::Arc;
+use tensorflow::library;
+use tensorflow::Tensor as TFTensor;
+use tensorflow::DEFAULT_SERVING_SIGNATURE_DEF_KEY;
+use tensorflow::{
+    FetchToken, Graph, Operation, SavedModelBundle, SessionOptions, SessionRunArgs, Status,
+};
+
+#[derive(Default)]
+pub(crate) struct TensorflowBackend();
+
+impl Backend for TensorflowBackend {
+    fn name(&self) -> &str {
+        "tensorflow"
+    }
+
+    fn load(
+        &mut self,
+        builders: &GraphBuilderArray<'_>,
+        _target: ExecutionTarget,
+        map_dir: &Option<Vec<(String, String)>>,
+    ) -> Result<Box<dyn BackendGraph>, BackendError> {
+        if map_dir.as_ref().is_some() {
+            if builders.len() != 2 {
+                return Err(BackendError::InvalidNumberOfBuilders(2, builders.len()).into());
+            }
+            // Initialize the Tensorflow backend
+            let retval = tensorflow::library::load().or_else(|e| {
+                println!("Error loading the TensorFlow backend: \n {}", e);
+                Err(e)
+            });
+
+            let unmap = map_dir.as_ref().unwrap();
+            let mut graph = Graph::new();
+            let dasbuilders = builders.as_ptr();
+            let export_dir = dasbuilders.read()?.as_slice()?;
+            let exp_str = str::from_utf8(&export_dir).unwrap();
+
+            for i in 0..unmap.len() {
+                if unmap[i].0 == exp_str {
+                    let bundle = SavedModelBundle::load(
+                        &SessionOptions::new(),
+                        &["serve"],
+                        &mut graph,
+                        Path::new(&unmap[i].1.clone()),
```

Comment:
Can this cause reading outside of dirs that the wasm module is given access to using a path like `../../../etc/passwd`? Wouldn't it be better to provide the model as a byte array just like for the openvino backend?

Reply:
The saved model API does not appear to provide a way to use bytes; ideally it would, though, so maybe it is worthwhile opening an issue in the tensorflow repository.

But @bjorn3's comment is a good one in that it highlights whether this is indeed correct. It doesn't seem so: a) shouldn't a user be allowed to `--map-dirs guest/foo:host/bar` and then ask wasi-nn to load the model at directory `guest/foo/my/model` (translated to `host/bar/my/model`? b) if this is the case, then this code will need to learn how to prevent escapes like @bjorn3 mentions, e.g., `../../../etc/password`.

I would propose that `map_dirs` be converted from `Vec<(String, String)>` to `Vec<MappedDir>` so that `MappedDir` can be the one place to hold this logic.